### ISSUE TITLE
[Surface Book 2] Suspend doesn't work

### ISSUE THREAD

**Issue by lukef** [jakeday/linux-surface#479]
Device: Surface Book 2
OS: Ubuntu 19.04

A quick note: If I can get a quick note on how to confirm my settings and how to diagnose/send you information, that would be awesome!

I'm seeing the following issues with suspend:

1. If I close the lid to put the device to sleep, the device will heat up. If it's inside a bag, the heat issue gets worse and after even 10 mins it will show the temperature warning when trying to wake it back up.
2. If the device is supended and it hasn't hit the temp warning then it just wont wake up. I have to hold the power button down until it shuts off and start it up again.

The heat issue is pretty worrying so I cannot use the suspend feature at all.

**Comment by qzed** [comment]
I'm definitely not an expert on suspend stuff as that has somehow mostly "just worked" for me, but I have a few things you could check:
- Which part of the device gets hot? I assume it's the CPU but on the off-chance that it's the dGPU, does installing https://github.com/qzed/linux-surfacebook2-mshw0153 help?
- Does this also occur when you suspend without closing the lid (e.g. by running `systemctl suspend`)? If not, have you installed anything else that might cause a suspend or hibernate on lid-close (e.g. acpid)? You could check that by running gnome-tweak-tools and disable "suspend when laptop lid is closed". I think that was the cause for a similar issue (#335) in the past.
- Do you have the `/lib/systemd/system-sleep/sleep` from this repo installed (could be something with IPTS)?

**Comment by lukef** [comment]
- I'm pretty sure it is the base that is getting hot and not the screen portion. They both get hot, but i think the base is heating up the screen because they're touching. I'll try installing `qzed/linux-surfacebook2-mshw0153`.
- I just left it unplugged for a while and it suspended (lid open). It still got really hot but i tapping power woke it back up. It seems to only become unrecoverable when the lid is closed.
- I have `/lib/systemd/system-sleep/sleep` installed.

**Comment by lukef** [comment]
If I disable the "suspend when laptop lid is closed" option in gnome-tweaks then closing the lid does nothing. The screen stays on and suspend isn't triggered at all.

**Comment by qzed** [comment]
Okay, then we can be sure it's not caused by multiple suspend calls. If you were able to wake it back up without closing the lid, can you get a dmesg log? I think that could help a lot.

**Comment by bruno222** [comment]
> The heat issue is pretty worrying so I cannot use the suspend feature at all.

Exactly the same behavior here on mine Surface Book 2. Will try to get some logs.

**Comment by buoyancy99** [comment]
Same here. Now I always turn the device off when close it. I am using arch on surface book 2 and in my case the issue involves not only heating but also freezing when I close the lid.

**Comment by qzed** [comment]
Sorry it took me this long to look at the logs, been a bit occupied. I can't identify a clear culprit here. There is a mention of `thermal_zone4` as being the one causing the shutdown, but I can't match that up with any sensor defined in ACPI as that name is not assigned deterministically. Due to the low threshold of 58°C it should be neither CPU nor GPU directly, but rather something like one of the batteries.

Apart from that, it looks like when this triggers, USB device access has issues, the EC doesn't react, and the nouveau driver crashes. This may of course all be caused by the system state going completely bonkers at that point.

Nevertheless (and since we haven't gotten a conclusive answer from @lukef yet): Does disabling the dGPU (e.g. via https://github.com/qzed/linux-surfacebook2-mshw0153 or UEFI) have any impact? Also what's your impression of where the heat comes from, GPU (base) or CPU (tablet)?

**Comment by lukef** [comment]
Sorry. I got busy. Will try to grab a log. As a side, I've noticed that when it suspends with the lid up it will still be slightly hot. I'm now convinced closing the lid just exacerbates the situation and it's not a unique situation.

**Comment by versipellis** [comment]
I have the heating issue with `systemctl suspend`. I can't quite tell yet whether it's the base heating up or the top heating up, since it takes a while to happen, but I have my dGPU disabled.

**Comment by qzed** [comment]
@versipellis Okay, since it's `systemctl suspend` it should not be #335 again.

**Comment by qzed** [comment]
It's odd that I still can't reproduce this. With the dGPU enabled I only get a slightly warm base, but that's the usual behavior, suspended or not. Nothing critical or even close to 58°C. How long does it usually take? Does this only happen off-power or connected to power?

**Comment by versipellis** [comment]
My usual process is to leave it plugged in most of the day, unplugged when going to meetings. I always disconnect it from my surface dock before suspending, just to be sure, then chuck it into my laptop case and head home. The reverse for going to work in the morning. All I know is when I take it out of my laptop case, it's warmer than when I put it in. It doesn't get exceedingly hot like in OP's case, although it used to do that in a previous kernel version, so I can't quite tell if it's built-up thermal from before I put it in the case that gets trapped in there, or if it's generating extra heat. I suspect it's the latter though, since it should have cooled down during my walk to work in New England winter ;)

**Comment by qzed** [comment]
@versipellis This might be two different issues then, still concerning though. Any chance the device wakes up at some point (e.g. I have to disable bluetooth if the pen is connected)?

**Comment by versipellis** [comment]
Nope, device doesn't wake. When it does, there's a loud hissing (same as the login one) that goes away the moment I touch the speaker audio output volume.

**Comment by qzed** [comment]
Hmm, okay.

**Comment by bruno222** [comment]
> Does disabling the dGPU (e.g. via https://github.com/qzed/linux-surfacebook2-mshw0153 or UEFI) have any impact? 

Testing... Will update you later on this.

> Also what's your impression of where the heat comes from, GPU (base) or CPU (tablet)?

The tablet behind the screen gets really, really hot.

> there's a loud hissing (same as the login one)

I have this as well (SB2), when I plug my headphones on the tablet (not on the dock)

**Comment by qzed** [comment]
One more thing you all could try is unloading the IPTS modules (as long as you don't need them). This could at least give us a hint if it's related to IPTS or rather the ACPI module. If the heat is coming from the tablet I'd guess it's one of those two.

**Comment by Resonious** [comment]
Hey, I'm having the exact same problem on Ubuntu 19.04. Adding the kernel parameter `module_blacklist=intel_ipts` fully fixes the problem for me.

**Comment by lukef** [comment]
I tried disabling `intel_ipts` (blacklist) and commenting it out from `/lib/systemd/system-sleep/sleep`. No change. Also, my laptop doesn't seem to sleep when closing the lid any more.

I'm now not 100% sure if it's actually going to sleep.

**Comment by qzed** [comment]
Still no idea what causes this, sorry. Doesn't seem IPTS at least.

**Comment by qzed** [comment]
@lukef Could be possible. What I've experienced is that bluetooth tends to wake up the device.

**Comment by runekaagaard** [comment]
Can confirm. @qzed fixed suspend at https://github.com/jakeday/linux-surface/pull/540. Tested on SB2 15", Ubuntu 19.04, Gnome. What a great time to be alive :) I made an ugly but functional workaround for the touchscreen rotation issue (wrong touchscreen coordinates when the screen is rotated) at https://github.com/runekaagaard/linux-surfacebook2-rotation-bugfix.
